mne/beamformer/_lcmv.py

```diff
@@ -65,29 +66,59 @@ def _setup_picks(picks, info, forward, noise_cov=None):
 
 
 @verbose
-def _apply_lcmv(data, info, tmin, forward, noise_cov, data_cov, reg,
-                label=None, picks=None, pick_ori=None, rank=None,
-                verbose=None):
+def _apply_lcmv(data, info, tmin, forward, reg, noise_cov=None,
+                data_cov=None, label=None, picks=None, pick_ori=None,
+                rank=None, weight_norm=None, verbose=None):
     """LCMV beamformer for evoked data, single epochs, and raw data."""
     is_free_ori, ch_names, proj, vertno, G = \
         _prepare_beamformer_input(info, forward, label, picks, pick_ori)
 
-    # Handle whitening + data covariance
-    whitener, _ = compute_whitener(noise_cov, info, picks, rank=rank)
-
-    # whiten the leadfield
-    G = np.dot(whitener, G)
-
-    # Apply SSPs + whitener to data covariance
+    # data covariance
+    if data_cov is None:
+        raise ValueError('Source reconstruction with beamformers requires '
+                         'a data covariance matrix.')
     data_cov = pick_channels_cov(data_cov, include=ch_names)
     Cm = data_cov['data']
+
+    # check number of sensor types present in the data
+    info_pick = pick_info(info, sel=picks)
+    ch_types =\
+        [_contains_ch_type(info_pick, tt) for tt in ('mag', 'grad', 'eeg')]
+    if sum(ch_types) > 1 and noise_cov is None:
+        raise ValueError('Source reconstruction with several sensor types '
+                         'requires a noise covariance matrix to be able to '
+                         'apply whitening.')
+
+    # apply SSPs
     if info['projs']:
         Cm = np.dot(proj, np.dot(Cm, proj.T))
-    Cm = np.dot(whitener, np.dot(Cm, whitener.T))
 
-    # Tikhonov regularization using reg parameter to control for
+    if noise_cov is not None:
+        # Handle whitening + data covariance
+        whitener, _ = compute_whitener(noise_cov, info, picks, rank=rank)
+        # whiten the leadfield
+        G = np.dot(whitener, G)
+        # whiten  data covariance
+        Cm = np.dot(whitener, np.dot(Cm, whitener.T))
+
+    # Tikhonov regularization using reg parameter d to control for
     # trade-off between spatial resolution and noise sensitivity
-    Cm_inv = _reg_pinv(Cm.copy(), reg)
+    Cm_inv, d = _reg_pinv(Cm.copy(), reg)
+
+    if weight_norm is not None:
+        # estimate noise level based on covariance matrix, taking the
+        # smallest eigenvalue that is not zero
+        noise, _ = linalg.eigh(Cm)
+        rank_Cm = estimate_rank(Cm, tol='auto', norm=False,
+                                return_singular=False)
+        noise = noise[len(noise) - rank_Cm]
+
+        # use either noise floor or regularization parameter d
+        noise = np.max([noise, d])
+
+        # Compute square of Cm_inv used for weight normalization
+        Cm_inv_sq = np.dot(Cm_inv, Cm_inv)
```

Comment:
It would be good to unify but I'm not sure what the consequences are. I suspect it would break a lot of things.

Do we use non-square in the MNE code?

We can also add a kwarg for it if we need to.